lib/galaxy_test/api/test_datasets.py

```diff
@@ -134,3 +134,30 @@ def test_tag_change(self):
         assert 'tag_c' in updated_hda['tags']
         assert 'name:tag_d' in updated_hda['tags']
         assert 'name:tag_e' in updated_hda['tags']
+
+    def test_update_datatype(self):
+        hda_id = self.dataset_populator.new_dataset(self.history_id)['id']
+        original_hda = self._get(
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id)).json()
+        assert original_hda['extension'] == 'txt'
+        assert original_hda['data_type'] == 'galaxy.datatypes.data.Text'
+        assert 'scatterplot' not in [viz['name'] for viz in original_hda['visualizations']]
+
+        update_while_incomplete_response = self._put(  # try updating datatype before upload is complete
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            {'datatype': 'tabular'})
+        self._assert_status_code_is(update_while_incomplete_response, 500)
```

Comment:
I think that could lead to unstable API tests as performance increases. We've got a a tool that can sleep before running, I'd use that for this test:
```
        inputs = {
            'input1': {'src': 'hda', 'id': hda2['id']},
            'sleep_time': 10,
        }
        run_response = self.dataset_populator.run_tool(
            "cat_data_and_sleep",
            inputs,
            history_id,
            assert_ok=False,
        )
        queued_id = run_response["outputs"][0]["id"]
```